File: Wonyoung/Programmers/92344.파괴되지 않은 건물/파괴되지않은건물.java
import java.util.*;

class Solution {   
    public int solution(int[][] board, int[][] skill) {
        int N = board.length;
        int M = board[0].length;
        
        int[][] prefix = new int[N+1][M+1];
        
        for(int[] turn : skill){
            int damage = turn[0] == 2 ? turn[5] : turn[5]*(-1);
            int r1 = turn[1], c1 = turn[2], r2 = turn[3], c2 = turn[4];
            
            prefix[r1][c1] += damage;
            prefix[r2 + 1][c1] -= damage;
            prefix[r1][c2 + 1] -= damage;
            prefix[r2 + 1][c2 + 1] += damage;
        }
        
        
        for(int i = 1; i <= N; i++){
            for(int j = 0; j <= M; j++){
                prefix[i][j] += prefix[i-1][j];
            }
        }
        
        for(int j = 1; j <= M; j++){
            for(int i = 0; i <= N; i++){
                prefix[i][j] += prefix[i][j - 1];
            }
        }
                
        
        int answer = 0;
        for(int i = 0; i < N; i++){
            for(int j = 0; j < M; j++){
                if(board[i][j] + prefix[i][j] > 0) answer++;
            }
        }
        return answer;
    }
    
}

File: Gwanghun/공통문제/9주차/프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.cpp
#include <string>
#include <vector>

using namespace std;

int diff[1001][1001];
//이를 통해 skill 사이즈만큼의 시간복잡도를 개선

int cal(vector<vector<int>> board, vector<vector<int>> skill) {
    
    int n = board.size();
    int m= board[0].size();
    
    for(auto s : skill){
        
        int type=s[0];
        
        int r1=s[1];
        int c1=s[2];
        int r2=s[3];
        int c2=s[4];
        
        int degree=s[5];
        
        int value=0;
        
        if(type==1) value = -degree;
        else value = degree;
        
        diff[r1][c1]+=value;
        diff[r1][c2+1]-=value;
        diff[r2+1][c1]-=value;
        diff[r2+1][c2+1]+=value;
        
    }
    
    for(int i=0; i<=n; i++){
        for(int j=1; j<=m; j++){
            diff[i][j]+=diff[i][j-1];
        }
    }
    
    for(int i=0; i<=m; i++){
        for(int j=1; j<=n; j++){
            diff[j][i]+=diff[j-1][i];
        }
    }
    
    int answer=0;
    
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            long long total = board[i][j]+diff[i][j];
            
            if(total>0) answer++;
        }
    }
    
    return answer;
}

int solution(vector<vector<int>> board, vector<vector<int>> skill) {
    int answer = cal(board,skill);
        
    return answer;
}
